### ISSUE TITLE
problems combining --substitute and label plugins

### ISSUE THREAD

**Issue by jmtd** [DBuildService/dock#94]
You can't use `--substitute` with  `pre_add_labels_in_df.py`, because the value for the substitution has to be a dictionary (`{labelname: labelvalue}`) and the substitution feature only supports string-values.

Sort-of related, labels should be reversed-domain according to [docker upstream](https://docs.docker.com/userguide/labels-custom-metadata/) (`com.example...` etc.), but the `--substitute` command uses the dot character to determine which part of the configuration to substitute within.

**Comment by jmtd** [comment]
Hi, related/masking this, the substitution has to be for a string only, whereas the `pre_add_labels_in_df` configuration needs a dict `{label: value}`

**Comment by jmtd** [comment]
I've worked around this for what I'm doing right now by writing a wrapper-plugin that accepts a flat string as an argument and decodes it into a JSON dictionary at initialisation (before handing the result to `pre_add_labels_in_df`: https://bitbucket.org/jboss-dockerfiles/dock/commits/4d2c2e462cf15ea4364544bd28df08f04d0307b8?at=jdowland-serialised-label-plugin

**Comment by TomasTomecek** [comment]
When this was implemented, we didn't even know about the reverse-domain policy.

`--substitute` option was implemented for you guys; so if you can come up with a better solution to this, I don't have problem merging it/improving it (but I feel like that passing json-encoded string via CLI is a bit rough -- on the other hand, we might not have another option)

**Comment by jmtd** [comment]
Yeah we're very grateful for `--substitute`, thank you! We make extensive use of it: I've only just extended our use to labelling (and we aren't using my extension in production yet, we're still evaluating it).

I considered submitting this

```
-- a/dock/plugin.py
+++ b/dock/plugin.py
@@ -340,7 +340,8 @@ class InputPlugin(Plugin):
                 build_json[key] = value
             else:
                 try:
-                    plugin_type, plugin_name, arg_name = key_fragments
+                    plugin_type, plugin_name = key_fragments[0],key_fragments[1]
+                    arg_name = '.'.join(key_fragments[2:])
                 except ValueError:
                     logger.error("invalid absolute path: it requires exactly three parts: "
                                  "plugin type, plugin name, argument name separated be dot")
```

But it didn't seem right as I wouldn't be running the affected code.

**Comment by TomasTomecek** [comment]
How about something like this:

```
--substitute "prebuild_plugins->add_labels_in_dockerfile->args->labels->com.example.key=value"
```

**Comment by goldmann** [comment]
@TomasTomecek Are you sure the example above is correct? Either case, why not use `{}` directly to specify the labels:

```
--substitute prebuild_plugins.add_labels_in_dockerfile.labels="{labelname: labelvalue, abc: 12}"
```

If it would be messy to make `--substitute` work with both - direct and non-direct substitutions we could add a marker to make it clear that the value is a valid json snippet:

```
--substitute json:prebuild_plugins.add_labels_in_dockerfile.labels="{labelname: labelvalue, abc: 12}"
```

or as part of the value (which could work better):

```
--substitute prebuild_plugins.add_labels_in_dockerfile.labels=json:"{labelname: labelvalue, abc: 12}"
```

**Comment by jmtd** [comment]
Hi, I took another look at this and PR #102 is an example of just overloading the existing plugin parameter. It didn't work out as messy as I thought it would, and shouldn't need any changes for other users of the plugin. What do you think?

**Comment by goldmann** [comment]
LGTM

**Comment by TomasTomecek** [comment]
If you guys are happy, I'm happy :) I really don't have strong feelings for one way or the other

Can this be closed since #102 is merged?

**Comment by goldmann** [comment]
@TomasTomecek Awesone, thank you! Yes, this issue can be closed now. If we'll need something else, we'll open a new one.
